File: LayoutTests/http/tests/plugins/resources/redirection-response.php
<?php
$status_code = $_GET['status'];

$uri = rtrim(dirname($_SERVER['PHP_SELF']), '/\\') . "/" . $_GET['target'];

$host = $_SERVER['HTTP_HOST'];
if (isset($_GET['host']))
    $host = $_GET['host'];

switch ($status_code) {
    case 301:
        header("HTTP/1.1 301 Moved Permanently");
        header("Location: http://" . $host . $uri);
        break;
    case 302:
        header("HTTP/1.1 302 Found");
        header("Location: http://" . $host . $uri);
        break;
    case 303:
        header("HTTP/1.1 303 See Other");
        header("Location: http://" . $host . $uri);
        break;
    case 307:
        header("HTTP/1.1 307 Temporary Redirect");
        header("Location: http://" . $host . $uri);
        break;
    default:
        header("HTTP/1.1 500 Internal Server Error");
        echo "Unexpected status code ($status_code) received.";
}
?>

File: LayoutTests/http/tests/plugins/resources/redirection-response.py
#!/usr/bin/env python3

import os
import sys
from urllib.parse import parse_qs

query = parse_qs(os.environ.get('QUERY_STRING', ''), keep_blank_values=True)
script_url = os.environ.get('SCRIPT_URL')
uri = '/'.join(script_url.split('/')[0:-1])
host = os.environ.get('HTTP_HOST')
status = int(query.get('status', [200])[0])
target = query.get('target', [''])[0]

sys.stdout.write('Content-Type: text/html\r\n\r\n')

if status in [301, 302, 303, 307]:
    sys.stdout.write(
        'status: {status}\r\n'
        'Location: http://{host}{uri}\r\n\r\n'.format(status=status, host=host, uri=uri)
    )
else:
    sys.stdout.write(
        'status: 500\r\n\r\n'
        'Unexpected status code ({}) received.'.format(status)
    )